api/pkg/controller/seed-sync/reconciler.go

```
@@ -0,0 +1,72 @@
+package seedsync
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+
+	kubermaticv1 "github.com/kubermatic/kubermatic/api/pkg/crd/kubermatic/v1"
+	"github.com/kubermatic/kubermatic/api/pkg/provider"
+	"github.com/kubermatic/kubermatic/api/pkg/resources/reconciling"
+
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type Reconciler struct {
+	ctrlruntimeclient.Client
+
+	seedKubeconfigGetter provider.SeedKubeconfigGetter
+	log                  *zap.SugaredLogger
+	ctx                  context.Context
+	recorder             record.EventRecorder
+}
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := r.log.With("seed", request.Name)
+	logger.Info("Reconciling seed")
+
+	seed := &kubermaticv1.Seed{}
+	if err := r.Get(r.ctx, request.NamespacedName, seed); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get seed: %v", err)
+	}
+
+	if err := r.reconcile(seed, logger); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to reconcile: %v", err)
+	}
+
+	logger.Info("Successfully reconciled")
+	return reconcile.Result{}, nil
+}
+
+func (r *Reconciler) reconcile(seed *kubermaticv1.Seed, logger *zap.SugaredLogger) error {
+	kubeconfig, err := r.seedKubeconfigGetter(seed)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve kubeconfig: %v", err)
+	}
+
+	client, err := ctrlruntimeclient.New(kubeconfig, ctrlruntimeclient.Options{})
+	if err != nil {
+		return fmt.Errorf("failed to create client for seed: %v", err)
+	}
+
+	namespaceCreators := []reconciling.NamedNamespaceCreatorGetter{
+		namespaceCreator(seed.Namespace),
```

Err, wouldn't the assumption be that the namesapce already exists, either because of helm or because of the kubermatic operator?

---

Inside the seeds? Isn't the seed possibly completely empty?

---

No, the kubermatic operator or helm is responsible for deploying kubermatic in it. This controller only runs after kubermatic got deployed in the master, its responsibility is to sync changes to the seed into the seed object in the seed cluster, not to install kubermatic there

---

Fair enough. Counterpoint: What's the harm in ensuring that the namespace exists?

---

Masking errors on a different layer, maintaining unneeded code

---

I very strongly disagree that adding these few lines creates a maintenance burden. What I can understand however is the argument that without Kubermatic already running inside the seed, the Seed CRD is not present and the reconciling would fail anyway. ;-) I still think that ensuring the namespace exists is a cheap and safe thing to do, both in terms of coding, maintaining and executing the code.